### PR TITLE
feat(tools): add use_ros - native ROS 2 bridge tool

### PR DESCRIPTION
**Pull request by cagataycali**
## What

Adds `use_ros`, a single agent-callable tool that bridges a Strands agent to any ROS 2 graph (topics + services) **entirely in-process through `rclpy`** - no `ros2` CLI shelling, no generated-code snippets. Every action calls the ROS 2 client library directly, so message types are real Python classes, errors are real exceptions, and a single long-lived node/executor is reused across calls.

```python
from strands import Agent
from strands_robots.tools import use_ros

agent = Agent(tools=[use_ros])
agent("list the ROS 2 topics, drive /turtle1 forward, and confirm its pose changed")
```

**Actions:** `status`, `list_topics`, `list_nodes`, `list_services`, `info`, `echo`, `publish`, `service_call`.

## Live showcase (real turtlesim, every action)

A reproducible, one-command harness drives a **real ROS 2 `turtlesim` node** through every action - in-process rclpy, including a closed **sense -> act -> sense** loop:

```bash
cd examples/ros2/use_ros_showcase
docker compose run --build --rm showcase   # exits 0 iff the turtle moved
```

Captured run (full transcript in `examples/ros2/use_ros_showcase/sample_output.txt`):

```
### use_ros(action='status')
[success] backend: rclpy (in-process)

### use_ros(action='list_nodes')
[success] /strands_robots_use_ros        <- our own node is on the graph
/turtlesim

### use_ros(action='info', topic='/turtle1/cmd_vel')
[success] topic info /turtle1/cmd_vel:
  type(s): geometry_msgs/msg/Twist
  publishers: 0
  subscribers: 1

### use_ros(action='echo', topic='/turtle1/pose', count=1)
[success] echo /turtle1/pose (turtlesim/msg/Pose):
[ { "x": 5.544, "y": 5.544, "theta": 0.0, ... } ]

### use_ros(action='publish', topic='/turtle1/cmd_vel',
             type='geometry_msgs/msg/Twist',
             fields={'linear': {'x': 2.0}, 'angular': {'z': 1.8}}, count=20)
[success] published 20 message(s) to /turtle1/cmd_vel

### use_ros(action='echo', topic='/turtle1/pose', count=1)   <- after driving
[success] [ { "x": 4.428, "y": 6.500, "theta": -1.445,
              "linear_velocity": 2.0, "angular_velocity": 1.8 } ]

### use_ros(action='service_call', service='/spawn',
             type='turtlesim/srv/Spawn', fields={'x':2.0,'y':2.0,'name':'t2'})
[success] response: { "name": "t2" }     <- /t2/* topics then appear in list_topics

### use_ros(action='echo', topic='/turtle1/pose', type='nonexistent_pkg/msg/Foo')
[error] use_ros: echo failed: No module named 'nonexistent_pkg'   <- structured, no crash

### use_ros(action='echo', topic='/bad; rm -rf')
[error] use_ros: invalid topic name: '/bad; rm -rf'               <- input validation
```

**The closed loop, proven:** a `geometry_msgs/msg/Twist` published through `use_ros` moved a real turtle from `(5.544, 5.544, 0.0)` to `(4.428, 6.500, -1.445)`, and the turtle's `linear_velocity`/`angular_velocity` latched to exactly the `2.0 / 1.8` commanded. Read back through `use_ros(action="echo")`. No `ros2` CLI anywhere.

## Design

- **In-process rclpy, single backend.** `rclpy` and `rosidl_runtime_py` must be importable - they ship with a sourced ROS 2 distro (apt / RoboStack / conda) and are **not** on PyPI, so they are not pinned in `pyproject.toml`. When absent, every action returns a clear, actionable error (it never raises); `use_ros(action="status")` reports `rclpy (in-process)` or `none`.
- **A single long-lived `Node` + `SingleThreadedExecutor`**, reused across calls and lock-guarded (rclpy spinning is not re-entrant). Graph introspection uses the rclpy node API directly (`get_topic_names_and_types`, `count_publishers`/`count_subscribers`, ...); pub/sub/service-call use real message classes resolved dynamically via `rosidl_runtime_py.get_message` / `get_service`, so any installed interface works with no static registry.
- **Field payloads are plain Python dicts** applied with `set_message_fields` and passed straight to rclpy - never serialised through source - so `bool`/`null` are preserved by construction.

## Safety

Agent-supplied topic/service/type names are validated against an allowlist before reaching the rclpy graph/type API (`_NAME_RE` for names, `_TYPE_RE` for `pkg/msg/Name` types). Because the tool never constructs a shell command or generates source, there is no command-injection or `eval` surface. Backend, timeout, and type-resolution failures return structured `{"status": "error"}` results rather than raising - including the `ModuleNotFoundError` raised when a valid-shaped type names a non-installed package (e.g. an LLM-hallucinated `nonexistent_pkg/msg/Foo`), which is caught by `except (ImportError, KeyError, AttributeError, ValueError, TypeError)` and pinned by `test_type_resolution_failure_is_structured`.

## Tests

`tests/tools/test_use_ros.py` runs ROS-free by monkeypatching the rclpy-facing helpers: every action-dispatch branch, input validation, the no-backend error path, the `fields` bool/null round-trip (dict reaches the writer intact), and the structured-error contracts (timeout, `ModuleNotFoundError`). `tests/tools/test_tools_lazy_import.py` updated for the lazy export. `ruff`, `ruff format`, and `mypy` clean.

## Packaging & docs

- Minimal `[ros2]` extra (`cyclonedds`; `rclpy`/`rosidl_runtime_py` are system packages not on PyPI, so the extra stays documented-and-minimal like `[curobo]`). Base install unchanged.
- `docs/ros2-integration.md` (in-process design, actions, safety, **Try it live**) + nav entry; `use_ros` rows in the README and `docs/hardware/tools.md`.
- `examples/ros2/use_ros_showcase/` - the reproducible live showcase above (Dockerfile + docker-compose + script + captured `sample_output.txt`).

## Scope

This PR is the core `use_ros` tool, complete and self-contained. Follow-ups (already filed, kept out to keep this a single unit):

- #740 - ROS 2 mesh bridge (`RosBridgedRobot`) - PR #742
- #741 - ROS 2 sim bridge (`SimEngine(ros2_bridge=True)`)

A complementary **pure-RTPS** backend (`use_rtps` / `RtpsRobot`, no rclpy, all ROS 2 distros, can *act as* a robot) builds on the same UX in #744.

## History

This tool was originally drafted with a two-backend design (native rclpy + a `docker exec` fallback, generating Python snippets run via `python3 -c`). It was rewritten to the in-process rclpy-only design above: the codegen approach forced a `json.dumps`->repr->`json.loads` round-trip, stdout scraping, and empty-except probes - all eliminated by calling rclpy directly. The `docker` backend and the `exec_raw` CLI escape hatch were dropped.

### CODE REVIEW

**Comment by cagataycali** [comment]
Follow-ups filed for the two bridge features that build on this tool (kept out of this PR to keep it a single self-contained unit):

- #740 - ROS 2 mesh bridge (`RosBridgedRobot` / `Robot.from_ros`)
- #741 - ROS 2 sim bridge (`SimEngine(ros2_bridge=True)`)

**Review by yinsong1986 — DISMISSED** [review]
## Summary

Adds `use_ros`, a single agent-callable tool that bridges a Strands agent to a ROS 2 graph (topics + services) with two auto-detected backends (`native` rclpy + `ros2` CLI, or `docker exec` into a sourced container) and a `none` fallback that returns a clear error. Types resolve dynamically through `rosidl_runtime_py`; JSON field payloads are applied with `set_message_fields`. Diff is self-contained: the tool, a thin lazy export, a minimal `[ros2]` extra, docs, and ROS-free behaviour tests.

I walked the security-sensitive paths since the whole surface interpolates agent-supplied strings into generated Python and subprocess argv. The defenses hold up:
- `topic`/`service` are regex-allowlisted (`_NAME_RE`) and `type` is `_TYPE_RE`-validated before reaching any helper.
- The `fields` dict is the one un-allowlisted agent input, but it is embedded as `json.dumps(fields)` then `{fields_json!r}` (a Python `repr` of a string) and parsed back with `json.loads` inside the helper. I verified a crafted `fields` value containing `' + __import__('os').system(...) + '` round-trips as inert data and cannot escape the repr'd string literal, so there is no code-injection vector.
- `exec_raw` rejects shell metacharacters and `shlex.split`s; native mode passes argv with no shell and docker mode `shlex.quote`s every token, so untrusted strings never reach a shell unescaped.
- An unrecognized `ROS2_MODE` override falls through to the `none` branch and returns a clean error dict rather than crashing.

The two CodeQL findings already on the PR (empty `except` at use_ros.py:97, dual import in the test) are non-blocking noise and already surfaced. No blocking concerns found.

## What's good

- Defense-in-depth input validation centralized at the tool entry point, with the repr+`json.loads` idiom correctly preserving `bool`/`null` (and a parametrized regression test pinning the bare-`true`/`false`/`null` `NameError` failure mode).
- ASCII-only tool output asserted in tests, consistent with the project's no-emoji rule.
- Scope discipline: mesh adapter and sim bridge explicitly deferred to follow-up PRs; base install unchanged; `[ros2]` extra kept minimal and documented like `[curobo]`.

## Verification suggestions

- Spot-check against `docker run -d --name ros-dev --net host ros:jazzy tail -f /dev/null` with `ROS2_MODE=docker`, then `use_ros(action="list_topics")` and a turtlesim publish/echo round-trip as described.
- Confirm `use_ros(action="exec_raw", command="topic list; echo pwned")` is rejected and an `echo` on a metachar topic returns the validation error.

File: strands_robots/tools/use_ros.py [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+    try:
+        out = subprocess.run(
+            ["docker", "inspect", "-f", "{{.State.Running}}", ROS2_DOCKER_CONTAINER],
+            capture_output=True,
+            text=True,
+            timeout=3,
+        )
+        if out.returncode == 0 and out.stdout.strip() == "true":
+            return "docker"
+    except (FileNotFoundError, OSError, subprocess.TimeoutExpired):
```
[the remark this answers is not in the document]
Comment (cagataycali) [reply strands_robots/tools/use_ros.py]: Addressed in 91e35ee. Both backend-probe except blocks now carry an explanatory comment for the intentional fall-through: the rclpy ImportError means the native backend is unavailable so the docker probe runs next, and the docker-inspect (FileNotFoundError/OSError/TimeoutExpired) case means no docker backend so detection returns 'none'. The swallow is deliberate (probe, not error path); behavior is unchanged.

File: tests/tools/test_use_ros.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+
+from __future__ import annotations
+
+import json
+import re
+from typing import Any
+
+import pytest
+
+import strands_robots.tools.use_ros as ros_mod
```
[the remark this answers is not in the document]
Comment (cagataycali) [reply tests/tools/test_use_ros.py:27]: Addressed in 91e35ee. The test needs the module object to monkeypatch internals (ros_mod._detect_mode, ros_mod._run_cli, etc.), so I kept `import ... as ros_mod` and replaced the second `from ... import use_ros` with a module-local alias `use_ros = ros_mod.use_ros`. The module is now imported exactly once; the 20 `use_ros(...)` call sites are unchanged and all 30 tests still pass.

**Review by yinsong1986 — DISMISSED** [review]
## Summary

Adds `use_ros`, a single agent-callable tool that bridges a Strands agent to any ROS 2 graph (topics + services) with two auto-detected backends (`native` rclpy + `ros2` CLI, or `docker exec` into a sourced container) and a `none` fallback that returns an actionable error. Nine actions (`status`, `list_topics`, `list_nodes`, `list_services`, `info`, `echo`, `publish`, `service_call`, `exec_raw`) dispatch through `_run_cli` / `_run_python`. Message/service types resolve dynamically via `rosidl_runtime_py`; JSON field payloads are applied with `set_message_fields`. The diff matches the description: +769/-0 across 9 files, self-contained tool plus packaging extra, docs, and tests.

## What's good

- **Input validation is centralized and applied before subprocess/snippet interpolation** — `_NAME_RE` / `_TYPE_RE` allowlists reject any agent-supplied topic/service/type carrying shell metacharacters or path traversal, and `exec_raw` additionally rejects `; | & $ \` < > \n \r \x00`. This is exactly the pattern AGENTS.md > Review Learnings (#92) > 'LLM Input Safety' prescribes.
- **Generated-helper injection surface is closed**: pre-validated names are interpolated with `!r` repr-quoting, and the `fields` payload round-trips through `json.dumps` + repr + `json.loads` rather than being pasted as Python source — never a code path. The parametrized regression in `test_use_ros.py` pins this contract.
- **Subprocess hygiene**: native mode passes argv with no shell; docker mode shlex-quotes every token.
- **Contract compliance**: all handlers return `{"status", "content"}` error dicts (never raise), ASCII-only output is asserted, the lazy export stays thin, and the two prior CodeQL findings (empty-except comment, dual import) are addressed at the current head.

## Verification suggestions

Standard CI (`hatch run test` over `tests/tools/`, ruff, mypy) covers this PR. For an optional live spot-check: `docker run -d --name ros-dev --net host ros:jazzy tail -f /dev/null && export ROS2_MODE=docker` then exercise `use_ros(action="list_topics")` and confirm `exec_raw` with a `;`-laden command is rejected.

File: strands_robots/tools/use_ros.py [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+                    return _err("service_call requires service and type")
+                import json
+
+                resp = _service_call(service, type, fields, timeout)
+                return _ok(f"response:\n{json.dumps(resp, indent=2, default=str)}")
+
+            return _err(f"unknown action: {action}")
+    except TimeoutError as exc:
+        return _err(str(exc))
+    except (KeyError, AttributeError, ValueError, TypeError) as exc:
```
Comment (yinsong1986) [inline strands_robots/tools/use_ros.py]: [MUST FIX] This except clause does not catch `ImportError`/`ModuleNotFoundError`, so a realistic agent input crashes the tool instead of returning a structured error.

`_get_message`/`_get_service` call `rosidl_runtime_py.utilities.get_message`, which calls `import_message_from_namespaced_type` -> `importlib.import_module('.'.join(message_type.namespaces))`. For a type string that passes `_TYPE_RE` but names a package that isn't installed (e.g. an LLM-hallucinated or misspelled `nonexistent_pkg/msg/Foo`), `importlib.import_module` raises `ModuleNotFoundError` (a subclass of `ImportError`). That is **not** in `(KeyError, AttributeError, ValueError, TypeError)`, so it propagates out of `use_ros` as an unhandled exception.

This violates AGENTS.md > Review Learnings (#85) > Error Handling Contracts: "Return error dicts, never raise - All AgentTool action handlers must return `{"status": "error", ...}` on failure. Never raise exceptions that bypass the structured response." `_TYPE_RE` only checks shape (`pkg/msg/Name`), not existence, so a valid-shaped but non-installed type is an ordinary, expected agent input on the `echo`/`publish`/`service_call` paths, not an exotic edge case.

Note the regression test `test_type_resolution_failure_is_structured` raises `KeyError` to simulate this, but the real failure mode is `ModuleNotFoundError`, so the test passes against code that still crashes in production (AGENTS.md > Testing > "Test import paths must match production").

Resolution: widen the clause to include `ImportError` (which covers `ModuleNotFoundError`), e.g. `except (ImportError, KeyError, AttributeError, ValueError, TypeError) as exc:`, and update the regression test to raise `ModuleNotFoundError` so it pins the real failure mode.

**Review by yinsong1986 — COMMENTED** [review]
## Summary

Adds `use_ros`, a single agent-callable Strands tool that bridges an agent to a ROS 2 graph entirely in-process through `rclpy` (no `ros2` CLI shelling, no code-generation). Eight actions are dispatched (`status`, `list_topics`, `list_nodes`, `list_services`, `info`, `echo`, `publish`, `service_call`), with dynamic message/service type resolution via `rosidl_runtime_py` and `set_message_fields` for payloads. A lazy `tools/__init__` export, a minimal `[ros2]` extra (cyclonedds), mypy overrides, README/docs/nav rows, and a ROS-free mocked test suite round it out.

Note: the PR description still describes the earlier two-backend design (native/docker, `exec_raw`, `ROS2_MODE`), but the shipped code is the in-process rclpy-only variant. That is a description/code drift, not a code blocker.

## What's good

- Input validation (`_NAME_RE` / `_TYPE_RE`) is applied up front before names reach the rclpy graph/type API, per AGENTS.md > LLM Input Safety.
- Concurrency contract is documented and the dispatch holds `_backend.lock`; no shell construction, so no command-injection surface.
- Tests run ROS-free via monkeypatched helpers, cover every dispatch branch, assert ASCII-only output, and pin the bool/None field round-trip.

## Verification suggestions

Against a sourced `ros:jazzy` + turtlesim, confirm `use_ros(action="publish", topic="/x", type="nonexistent_pkg/msg/Foo")` returns a `{"status": "error"}` dict rather than raising (see inline comment).

File: strands_robots/tools/use_ros.py (hunk 3 above)
Comment (cagataycali) [reply strands_robots/tools/use_ros.py]: Fixed in a532aa6. You are right: a valid-shaped type that names a non-installed package resolves through get_message -> import_message_from_namespaced_type -> importlib.import_module, which raises ModuleNotFoundError (a subclass of ImportError), and that was not in the except tuple, so it escaped as an unhandled exception rather than a structured error.

Resolution applied exactly as suggested:
- Widened the clause to `except (ImportError, KeyError, AttributeError, ValueError, TypeError)` so ModuleNotFoundError is caught and surfaced as `{"status": "error"}`.
- Updated `test_type_resolution_failure_is_structured` to raise `ModuleNotFoundError` (the real failure mode) instead of `KeyError`. Verified the test fails against the pre-fix clause (ModuleNotFoundError propagates unhandled) and passes after the widen, so the regression now pins the actual production crash rather than a simulated one.

Comment (cagataycali) [reply strands_robots/tools/use_ros.py]: Fixed in `a532aa6`. The except clause now reads `(ImportError, KeyError, AttributeError, ValueError, TypeError)`, and `test_type_resolution_failure_is_structured` raises `ModuleNotFoundError` (the real production failure mode) instead of `KeyError`.

Pin test: `tests/tools/test_use_ros.py::test_type_resolution_failure_is_structured`

---
<sub>This response was generated by an AI agent as part of the strands-robots autonomous contributor workflow.</sub>

**Comment by cagataycali** [comment]
### Showcase: a Strands agent drives a real turtlesim through `use_ros`

Beyond calling the tool directly, I wired `use_ros` to a Bedrock-backed Strands agent (`global.anthropic.claude-opus-4-8`) and asked it, in plain English, to **"draw a square"**. With no further guidance it sequenced **9 `use_ros` tool calls** - 4 forward sides, 3 left turns, a stop, and an echo - driving a real `turtlesim` node (in-process rclpy, no `ros2` CLI), then read back and reported the final pose:

```
pose BEFORE: (5.54, 5.54, 0.0)
=== AGENT DRIVING (square) ===
Side 1: Tool #1: use_ros   Turn 1: Tool #2: use_ros   ...   Stop: Tool #8   Echo: Tool #9
"Square complete! Final pose: x=5.35, y=6.73, theta~2.84 rad, velocities at zero.
The path returned near the start area, confirming a roughly closed square."
pose AFTER: (5.35, 6.73, 2.84)
```

The live turtlesim canvas was recorded (Xvfb + ffmpeg) while the agent drove - see `docs/assets/use_ros_agent_turtle.gif`/`.mp4`.

Reproducible artifacts added under `examples/ros2/use_ros_showcase/`:
- `agent_drive.py` - the agent driver (Bedrock + `use_ros`, natural-language task)
- `showcase.py` - direct exercise of every action (+ one-command `docker compose run --build --rm showcase`)
- `sample_output.txt` / `agent_sample_output.txt` - captured live runs

This exercises the full action surface against a real ROS 2 graph - `status`, `list_topics`/`nodes`/`services`, `info` (live pub/sub counts), `echo`, `publish`, `service_call` (spawns a second turtle) - plus both structured-error contracts: a non-installed type (`nonexistent_pkg/msg/Foo`) returns `{"status":"error"}` rather than crashing (the `ModuleNotFoundError` fix from a532aa6), and a metacharacter topic is rejected by validation.

**Comment by cagataycali** [comment]
Resolved the CodeQL `py/call-to-non-callable` error.

Root cause: `agent_drive.py` imported the tool via the package re-export (`from strands_robots.tools import use_ros`). `strands_robots.tools` re-exports tools lazily through module `__getattr__`, and the package also ships a like-named submodule (`strands_robots/tools/use_ros.py`). Static analysis can't see the lazy re-export, so it resolves `use_ros` to the submodule object; the subsequent `use_ros(...)` call then reads as calling a module (non-callable).

Fix: import the tool from its concrete module (`from strands_robots.tools.use_ros import use_ros`), matching the sibling `showcase.py`, so the name unambiguously resolves to the callable tool. Added a regression test that scans the showcase scripts: any script calling `use_ros()` directly must import it from its concrete module, not the package re-export (fails on the pre-fix import). Local `ruff`/`ruff format`/`mypy` clean; full `use_ros` suite passes.

**Comment by cagataycali** [comment]
Closing in favor of #746.

The `use_ros` tool itself already landed on `main` via #742 (the `RosBridgedRobot` mesh bridge was built on it and carried the tool in), so this PR's tool diff is now empty against `main`. What remained was the example/showcase, which #746 delivers cleanly from a fresh `main` base (no conflicts), relocated to `examples/ros2/use_ros/` and with a new agent-driven closed-loop square video.

#746 also carries the one outstanding review fix from here - widening the dispatch except to catch `ModuleNotFoundError` (the real failure mode when a valid-shaped type names a non-installed package), with its regression test.

Thanks @yinsong1986 for the reviews on this branch - the fixes carried forward.

**Comment by Vivek0712** [comment]
`strands_robots/tools/use_ros.py`
 
### `echo` cannot receive sensor-data topics on real hardware (QoS mismatch)
 
`_echo` subscribes with a bare depth (`node.create_subscription(msg_cls, topic, cb, 10)`), which yields the default RELIABLE QoS profile. On real robots, `/scan`, camera, and most sensor topics publish BEST_EFFORT (SensorDataQoS). A RELIABLE subscription is incompatible with a BEST_EFFORT publisher, so the subscription never matches, `spin_for` runs out the timeout, and the tool returns a *success* result with zero samples.
 
Tested on a TurtleBot4: `use_ros(action="echo", topic="/scan")` times out every time, while `ros2 topic echo /scan --qos-profile sensor_data` works. This silently breaks `RosBridgedRobot.get_scan()` too, since it forwards here.
 
Suggested fix: query `node.get_publishers_info_by_topic(topic)` and adopt the publisher's reliability/durability (rclpy's adaptive-QoS idiom), or at minimum add a `qos: str = "default" | "sensor_data"` parameter and use `rclpy.qos.qos_profile_sensor_data`.

**Comment by Vivek0712** [comment]
###  Multi-message `publish` holds the global backend lock for its full duration - a long `drive` is uninterruptible and `stop` is blocked
 
`use_ros` wraps every action in `with _backend.lock:`, and `_publish` sleeps inside that lock for `count/rate` seconds. `RosBridgedRobot.drive(duration=10)` therefore holds the process-wide ROS lock for 10 s: any concurrent `use_ros` call - including `stop()` publishing a zero Twist, or `get_scan()` checking for an obstacle - blocks until the burst completes. The showcase prompt "drive forward, stop if /scan sees obstacle <0.5m" cannot actually preempt a long drive under these semantics; the agent can only approximate it by interleaving many short drives with scans.
 
Suggested direction: either (a) release the lock between publishes (create the publisher under the lock, publish/sleep outside it - rclpy publishers are thread-safe for publish), or (b) document that `duration` should stay short and preemption is achieved by chunking. Worth deciding before the interface calcifies.
